sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -439,6 +439,39 @@ object StateStoreProvider {
   }
 }
 
+/**
+ * This is an optional trait to be implemented by [[StateStoreProvider]]s that can read fine
+ * grained state data which is replayed from a specific snapshot version. It is used by the
+ * snapshotStartBatchId option in state data source.
+ */
+trait SupportsFineGrainedReplayFromSnapshot {
+  /**
+   * Used by snapshotStartBatchId option when reading state generated by join operation as data
```

Comment:
Let's not couple too much with implementation details, especially the current implementation of Spark codebase. 3rd party state store provider does not need to know about this. If we think this is needed to please them to implement the two different methods, let's just leave this method for the interface level and wrap the state store with read-only in caller side.